Ruby: Add support for proto3 json_name in compiler and field definitions

lfittl opened the pull request:
This adds support for the `json_name` field definition, per the Proto3 spec.

Note the JSON output mechanism is already reading the `json_name` from ubp's struct (see [here](https://github.com/protocolbuffers/protobuf/blob/cf28f28347d65e352f429263500f982ae3e51aed/ruby/ext/google/protobuf_c/ruby-upb.c#L5730)), but this was not being added to the generated Ruby DSL, or read in from the DSL into the ubp field definitions.

lfittl commented:
@googlebot I signed it!

lfittl commented:
@haberman I assume you would be the correct person to review this further?

(Thanks for your work on the Ruby protobuf support!)

lfittl commented:
@haberman Ping on this, tagging you since you seem to be the primary maintainer for the Ruby bindings - not sure if I missed anything here. Appreciate the help - thank you :)

haberman commented:
Hi @lfittl , I'm sorry for my slow reply on this.

I'm somewhat reluctant to go farther down this path of extending the DSL, because it is an un-winnable game of whack-a-mole. The DSL fundamentally can't represent all options, particular custom options that are used in their own definition:

```
syntax = "proto3";

import "google/protobuf/descriptor.proto";

extend google.protobuf.FieldOptions {
  optional int32 foo_option = 1000 [(foo_option) = 5];
}
```

That said, the fact that `json_name` is a member of `FieldDescriptorProto` (and not `FieldOptions`) does make the PR more compatible with a future world where we do handle options correctly. So maybe we can move forward with this now.

lfittl commented on ruby/ext/google/protobuf_c/defs.c:
```diff
@@ -1750,6 +1767,19 @@ static void msgdef_add_field(VALUE msgbuilder_rb, upb_label_t label, VALUE name,
+          rb_hash_lookup(options, ID2SYM(rb_intern("json_name")));
+
+      /* Call #to_s since json_name is always a string in the descriptor. */
+      json_name = rb_funcall(json_name, rb_intern("to_s"), 0);
```
Yup, you are correct - this was copied from the default value handling above, but `json_name` is indeed always a string, so this conversion is unnecessary.

lfittl commented:
> Hi @lfittl , I'm sorry for my slow reply on this.

No worries, I know the same from my own open source projects, thanks for taking a look!

> 
> I'm somewhat reluctant to go farther down this path of extending the DSL, because it is an un-winnable game of whack-a-mole. The DSL fundamentally can't represent all options, particular custom options that are used in their own definition:
> 
> ```
> syntax = "proto3";
> 
> import "google/protobuf/descriptor.proto";
> 
> extend google.protobuf.FieldOptions {
>   optional int32 foo_option = 1000 [(foo_option) = 5];
> }
> ```
> 
> That said, the fact that `json_name` is a member of `FieldDescriptorProto` (and not `FieldOptions`) does make the PR more compatible with a future world where we do handle options correctly. So maybe we can move forward with this now.

Yeah, I see how it's important to handle this generically at some point - but I think it still makes sense to do this now, since it allows use of `json_name` which seems to me one of the core features of Proto3 (as opposed to other custom field tags).

---

I've addressed the review comments you shared, and also added a test that covers this, to ensure it doesn't break by accident in the future.
